sled-agent/src/bootstrap/server.rs

```diff
@@ -210,15 +266,57 @@ async fn serve_single_request(
     let response = match envelope.request {
         Request::SledAgentRequest(request) => {
             match bootstrap_agent.request_agent(&*request).await {
-                Ok(response) => Ok(Response::SledAgentResponse(response)),
+                Ok(response) => {
+                    // If this request succeeded, we should update our list of
+                    // trust quorum members to match the request. In the common
+                    // case, we should only receive (and successfully handle) a
+                    // sled agent request if we are starting up for the first
+                    // time, meaning we won't have a list of trust quorum
+                    // members (we just received it in `request`!). Less
+                    // commonly, we may receive a duplicate sled agent request.
+                    // However, we should never successfully handle a sled agent
+                    // request that contains a different trust quorum share
+                    // (include the set of members); we assert to confirm that
+                    // in the non-`None` case.
+                    if current_trust_quorum_members.is_none() {
+                        *trust_quorum_members.lock().unwrap() = request
+                            .trust_quorum_share
+                            .as_ref()
+                            .map(|dist| dist.member_device_id_certs.clone());
+                    } else {
+                        assert!(
+                            current_trust_quorum_members.as_ref()
+                                == request
+                                    .trust_quorum_share
+                                    .as_ref()
+                                    .map(|share| &share.member_device_id_certs),
+                            concat!(
+                                "Unexpectedly succeeded handling a sled-agent",
+                                "request with different trust quorum members",
+                            )
+                        );
+                    }
+                    Ok(Response::SledAgentResponse(response))
+                }
                 Err(err) => {
                     warn!(log, "Sled agent request failed"; "err" => %err);
                     Err(format!("Sled agent request failed: {err}"))
                 }
             }
         }
         Request::ShareRequest => match bootstrap_agent.secret_share().await {
-            Some(share) => Ok(Response::ShareResponse(share)),
+            Some(dist) => {
+                // We should never return a secret share without having verified
+                // the peer is a member of our trust quorum.
+                // `bootstrap_agent.secret_share()` should be `None` if we don't
+                // know our peers (implying we haven't yet received our initial
+                // sled-agent request); assert here to be sure.
+                assert!(
```

Comment:
Sorry, I'm not sure I follow - we shouldn't be waiting for the trust quorum to handle this request. Where does it look like it is?

If we get a share request before we've received our share from RSS, `bootstrap_agent.secret_share()` should return `None`, which would cause us to return an error to the client. That said, I think there may be a race here if we get both the share from RSS and a client request for the share simultaneously; need to reread to see.

Reply:
Ok, reworked this in b5a60b07:

* Inner now holds the `ShareDistribution` instead of `BootstrapAgent` (BA still holds the share to compare against duplicate share requests, but that's its only use of it)
* Inner has separate `accept()` loops for pre- and post-initialization; the former knows the quorum isn't available yet and only responds to init requests; the latter knows the quorum is available and only responses to share retrieval requests

Reply:
Thanks for the rework. I really like both these ideas. And yes, as discussed, my question above was from an under-caffeinated reading of the code.